ring_theory/localization.lean

```diff
@@ -95,32 +97,147 @@ by refine
+@[simp] lemma of_comm_ring_zero : (of_comm_ring 0 : loc α S) = 0 := rfl
+@[simp] lemma of_comm_ring_one : (of_comm_ring 1 : loc α S) = 1 := rfl
+@[simp] lemma of_comm_ring_add {x y : α} :
+  (of_comm_ring (x + y) : loc α S) = of_comm_ring x + of_comm_ring y := is_ring_hom.map_add _
+@[simp] lemma of_comm_ring_sub {x y : α} :
+  (of_comm_ring (x - y) : loc α S) = of_comm_ring x - of_comm_ring y := is_ring_hom.map_sub _
+@[simp] lemma of_comm_ring_mul {x y : α} :
+  (of_comm_ring (x * y) : loc α S) = of_comm_ring x * of_comm_ring y := is_ring_hom.map_mul _
+@[simp] lemma of_comm_ring_neg {x : α} :
+  (of_comm_ring (-x) : loc α S) = -of_comm_ring x := is_ring_hom.map_neg _
+@[simp] lemma of_comm_ring_pow {x : α} {n : ℕ} :
+  (of_comm_ring (x ^ n) : loc α S) = of_comm_ring x ^ n := is_semiring_hom.map_pow _ _ _
+
+instance : has_coe α (loc α S) := ⟨of_comm_ring⟩
+
+instance coe.is_ring_hom : is_ring_hom (coe : α → loc α S) :=
+localization.of_comm_ring.is_ring_hom
+
+@[simp] lemma coe_zero : ((0 : α) : loc α S) = 0 := rfl
+@[simp] lemma coe_one : ((1 : α) : loc α S) = 1 := rfl
+@[simp] lemma coe_add {x y : α} : (↑(x + y) : loc α S) = x + y := of_comm_ring_add
+@[simp] lemma coe_sub {x y : α} : (↑(x - y) : loc α S) = x - y := of_comm_ring_sub
+@[simp] lemma coe_mul {x y : α} : (↑(x * y) : loc α S) = x * y := of_comm_ring_mul
+@[simp] lemma coe_neg {x : α} : (↑(-x) : loc α S) = -x := of_comm_ring_neg
+@[simp] lemma coe_pow {x : α} {n : ℕ} : (↑(x ^ n) : loc α S) = x ^ n := of_comm_ring_pow
+
+@[simp] protected lemma div_self {x : α} {hx : x ∈ S} :
+  (div x ⟨x, hx⟩ : loc α S) = 1 :=
+quotient.sound ⟨1, is_submonoid.one_mem S,
+by simp only [subtype.coe_mk, is_submonoid.coe_one, mul_one, one_mul, sub_self]⟩
+
+@[simp] lemma div_self' {s : S} :
+  (div s s : loc α S) = 1 :=
+by cases s; exact localization.div_self
+
+@[simp] lemma div_self'' {s : S} :
+  (div s.1 s : loc α S) = 1 :=
+div_self'
+
+@[simp] lemma coe_mul_div (x y : α) (s : S) :
+  ↑x * div y s = div (x * y) s :=
+quotient.sound $ r_of_eq $ by rw one_mul
+
+protected lemma div_eq_mul_one_div (r : α) (s : S) :
+  div r s = r * div 1 s :=
+by rw [coe_mul_div, mul_one]
+
+@[simp] protected lemma div_mul_div (x y : α) (s t : S) :
+  div x s * div y t = div (x * y) (s * t) := rfl
+
+@[simp] protected lemma mul_div_cancel_left (r : α) (s : S) :
+  div (↑s * r) s = r :=
+by rw [localization.div_eq_mul_one_div, mul_comm ↑s, coe_mul,
+       mul_assoc, ← localization.div_eq_mul_one_div, div_self', mul_one]
+
+@[simp] lemma mul_div_cancel_right (r : α) (s : S) :
+  div (r * s) s = r :=
+by rw [mul_comm, localization.mul_div_cancel_left]
+
+@[elab_as_eliminator]
+protected theorem induction_on {C : loc α S → Prop} (x : loc α S)
+  (ih : ∀ r s, C (div r s : loc α S)) : C x :=
+by rcases x with ⟨r, s⟩; exact ih r s
+
+@[elab_with_expected_type]
+protected def rec {β : Type v} [comm_ring β]
+  (f : α → β) (hf : is_ring_hom f) (g : S → units β) (hg : ∀ s, (g s : β) = f s)
```

Comment:
Wouldn't it make more sense to replace `g` and `hg` with `(hS : ∀ s, is_unit (f s))`?
(This needs `ring_theory/associated.lean`.)

Reply:
This is done to preserve computability.

Reply:
done